clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapServerStore.java

```diff
@@ -226,10 +144,51 @@ private void writeUnlockAll() {
     }
   }
 
-  boolean handleOversizeMappingException(long hash) {
+  private void handleOversizeMappingException(long key, LongConsumer operation) {
+    handleOversizeMappingException(key, k -> {
+      operation.accept(k);
+      return null;
+    });
+  }
+
+  /**
+   * Force eviction from other segments until {@code operation} succeeds or no further eviction is possible.
+   *
+   * @param key the target key
+   * @param operation the previously failed operation
+   * @param <R> operation result type
+   * @return the operation result
+   * @throws OversizeMappingException if the operation cannot be made to succeed
+   */
+  private <R> R handleOversizeMappingException(long key, LongFunction<R> operation) throws OversizeMappingException {
+    if (tryShrinkOthers(key)) {
+      try {
+        return operation.apply(key);
+      } catch (OversizeMappingException ex) {
+        //ignore
+      }
+    }
+
+    writeLockAll();
+    try {
+      OversizeMappingException e;
+      do {
+        try {
+          return operation.apply(key);
+        } catch (OversizeMappingException ex) {
+          e = ex;
+        }
+      } while (tryShrinkOthers(key));
+      throw e;
```

Comment:
I still don't get with we are throwing `e`. In the end, if the `operation.apply` worked, aren't we happy? Also, if no amount of shrinking can safe us, aren't we in an infinite loop? Finally, `tryShrinkOthers` try to shrink every other segments. If it fails the first time, isn't it meaning we are doomed already? And so throw an `OversizeMappingException`? No need to have a while loop. Some line comments can possible help me understand.

Reply:
If the `operation.apply(key)` worked then the method will have returned.  We'll only get to `throw e` if we failed to shrink other segments. We've shrunk as much as possible and we still can't complete the operation.  Make sense?

Reply:
Yes... I think I missed the return.